### PR TITLE
SIMD-0088: Enable Core BPF Programs

### PR DESCRIPTION
**Pull request by buffalojoec**
This proposal defines the concept of "Core BPF Programs" - programs which are essential to cluster operations but exist as BPF programs, rather than native built-in programs.

The proposal describes, in a general sense, how native programs are migrated to Core BPF, and defines the requirements for proposing a native program migration to Core BPF.

### CODE REVIEW

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [inline proposals/0088-enable-core-bpf-programs.md:46]: This is going outside of the motivation and summarizing the approach. How about moving this to the summary?

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [inline proposals/0088-enable-core-bpf-programs.md:50]: nit: let's use the new terminology!
```suggestion
An alternative approach to managing upgrades to core BPF programs could be
```

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [inline proposals/0088-enable-core-bpf-programs.md:56]: Sorry, I don't think I understand this point. How does a feature gate come into play if you're using a multisig? 

From the way I understand it, a multisig is more dangerous because it means potential changes at any time, whereas a feature gate is preferred because it has such a long lead time to deployment, and requires validators to upgrade their software to signal their acceptance of the change.

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [inline proposals/0088-enable-core-bpf-programs.md:71]: Side note: we should check with the runtime team to see if it's preferred to have them as "final" upgradeable programs. I always forget which one is better for the runtime and avoids copying the program data into the vm

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [inline proposals/0088-enable-core-bpf-programs.md:66]: This doesn't seem like new terminology, and more about defining terms for the design. Let's move these into the detailed design portion

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [inline proposals/0088-enable-core-bpf-programs.md:76]: nit: it's not the source code deployed
```suggestion
to a new arbitrary address as a **non-upgradeable** BPF program and
```

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [inline proposals/0088-enable-core-bpf-programs.md:78]: nit: It might be worth making it completely clear what the runtime needs to do, ie
```suggestion
using a feature gate to move the modified program in
place of the existing program. The feature gate is required to circumvent
normal transaction processing rules and replace the contents of one account
with another directly at the runtime level.
```

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [inline proposals/0088-enable-core-bpf-programs.md:107]: Sorry, what does this mean?

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [inline proposals/0088-enable-core-bpf-programs.md:100]: Up to you, but I think there might be more to this since the runtime will also need to remove any special-casing for that address during transaction processing.

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [inline proposals/0088-enable-core-bpf-programs.md:119]: I think we can remove this line

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [inline proposals/0088-enable-core-bpf-programs.md:139]: It might be nice to point out that the core BPF implementation must be more than backwards compatible, and provide exactly the same results as the current native program.

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [inline proposals/0088-enable-core-bpf-programs.md:136]: I'm not sure this is true, since the BPF program will exist at the address instead of the native program, which is a breaking change on the network.

**Review by joncinque — COMMENTED** [review]
Looks really good for the most part! Mostly some small bits to clean up, then we can start getting more eyes on it

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:136]: Oh, I was referring to the addition of the code that does the swap/upgrade. Without actually using it on a program, merely introducing it off the hot path, doesn't introduce any breaking changes.

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:100]: I think it's worth including the additional steps in the SIMD. What does it look like to remove special casing?

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:107]: I was originally moving toward some validation/checks that had to do with native program instruction/state, but I don't think it's going to work or worth adding.

This was a straggler from those ideas.

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:72]: Asked the VM guys, will circle back with an update.

File: proposals/0088-enable-core-bpf-programs.md
Comment (joncinque) [reply proposals/0088-enable-core-bpf-programs.md]: Oh I see, so you have the option of doing either upgrade path? That seems overly complicated to describe, and we'll get much more out of differentiating with a plain multisig

Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md]: Lmk if the updated description is too verbose.

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:72]: Actually circling back, it appears neither has a significant VM-related advantage.

With that being said, I feel like, since v2 (non-upgradeable) is _not_ deprecated, it actually feels a little more official to lock these special Core BPF programs into the "final" loader - as written in the SIMD now. Basically, they're intentionally designed to not be upgraded by a keypair authority. What do you think?

File: proposals/0088-enable-core-bpf-programs.md
Comment (CriesofCarrots) [inline proposals/0088-enable-core-bpf-programs.md:111]: This makes me feel more strongly about removing the phrase "modified program" everywhere above.

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:71]: ahh whoops!

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md]: Just for clarity, the concept of "which loader owns all formerly native Core BPF programs" has been removed from this SIMD. Now, one must define this specification in the SIMD that outlines that specific native program's migration to Core BPF.

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md]: Since this SIMD is now slimmed down and mentions that upon any migration, a SIMD outlining the migration and Core BPF specifications should be published, I think describing special casing removal isn't necessary here.

However, I've added it to the necessary details for such a SIMD as mentioned in this proposal.

**Comment by buffalojoec** [comment]
@joncinque @CriesofCarrots I think this SIMD is good to go.

@lheeger-jump @ripatel-fd can you guys have a look please?

File: proposals/0088-enable-core-bpf-programs.md
Comment (CriesofCarrots) [inline proposals/0088-enable-core-bpf-programs.md:19]:
```suggestion
This SIMD details the process by which existing native programs can be
```

File: proposals/0088-enable-core-bpf-programs.md
Comment (CriesofCarrots) [inline proposals/0088-enable-core-bpf-programs.md:25]:
```suggestion
BPF programs offer less complexity than native programs for other clients, such as Firedancer, since
```

File: proposals/0088-enable-core-bpf-programs.md
Comment (CriesofCarrots) [inline proposals/0088-enable-core-bpf-programs.md:62]:
```suggestion
BPF version of the native program to a new arbitrary address and using a
feature gate to move the BPF program, replacing the existing native program at
```

**Review by CriesofCarrots — COMMENTED** [review]
Just a couple language suggestions. Content lgtm!

**Review by CriesofCarrots — APPROVED** [review]

File: proposals/0088-enable-core-bpf-programs.md
Comment (2501babe) [inline proposals/0088-enable-core-bpf-programs.md:77]: do you have a mechanism in mind for how this replacement would take place? unless im missing something even with loader v4 you would need to run a transaction to replace the bytecode of target with source

**Comment by 2501babe** [comment]
just hopping in because ill be working on some of these ports, ty for proposing a process!

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:74]: I'm thinking it _should_ work if you're just deploying and not invoking the program (right?).

In theory, if the core BPF program at say `11111111111111111111111111111111` has special privileges, then unless the executable lives at `11111111111111111111111111111111` it's not going to be able to use those privileges. So you deploy the BPF version of it to some arbitrary address, and if it was to be invoked at that arbitrary address it would fail to load. However, it shouldn't ever be invoked at that arbitrary address. It's only temporarily living there until the swap is complete.

Additionally, I think if any tweaks are necessary in the future as we figure out how to provide all of these core BPF programs with the necessary privileges, we can issue another proposal to modify the process or introduce a new one.

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:77]: We would basically build out a revamped version of this function, and call it from a runtime feature gate on epoch rollover.

https://github.com/solana-labs/solana/blob/a2d2eb3ba80824f55c9cbd267ab881a285ca324a/runtime/src/bank.rs#L7985

**Review by joncinque — APPROVED** [review]

File: proposals/0088-enable-core-bpf-programs.md
Comment (lheeger-jump) [inline proposals/0088-enable-core-bpf-programs.md:59]: This show be standardized

File: proposals/0088-enable-core-bpf-programs.md
Comment (lheeger-jump) [inline proposals/0088-enable-core-bpf-programs.md:57]:
```suggestion
- How any required special privileges will be granted to the program in its BPF
- The interface for the program.
- A precise and complete specification of the behavior.
```

File: proposals/0088-enable-core-bpf-programs.md
Comment (lheeger-jump) [reply proposals/0088-enable-core-bpf-programs.md:74]: More discussion is required from VM stakeholders about privileged execution before I can consider approval. I do not want to add more syscalls if they all have to be re-invented in PRv2.

Comment (lheeger-jump) [inline proposals/0088-enable-core-bpf-programs.md:74]: Can't we just use the same address?

**Review by lheeger-jump — CHANGES_REQUESTED** [review]
I also apologize for my very late review. We need more input from the VM team and the deployment procedure is unclear. Would this be an EOA that controls the account? I think changing the address would really hurt app devs and UX.

File: proposals/0088-enable-core-bpf-programs.md
Comment (Lichtso) [reply proposals/0088-enable-core-bpf-programs.md:74]: I think we need to replace the program at the original address anyway, otherwise all its owned accounts will not be available to the replacement.

Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:74]: To replace a native program with a BPF program, you'd have to deploy the BPF version to some arbitrary address, then move it into the account at the original address. That would result in the address being unchanged and the program being replaced with BPF.

See [my comment here](https://github.com/solana-foundation/solana-improvement-documents/pull/88#discussion_r1441929536).

Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:74]: > More discussion is required from VM stakeholders about privileged execution before I can consider approval. I do not want to add more syscalls if they all have to be re-invented in PRv2.

Fair point. I guess the question is, does the number of syscalls block us from migration _any_ native program to BPF?

In other words, is it possible we could move this SIMD along to acknowledge that at least _some_ programs will be migrated, and then assess the impact of any necessary syscalls when the SIMDs outlining migration specifications are proposed for a program requiring privileges?

Comment (lheeger-jump) [reply proposals/0088-enable-core-bpf-programs.md:74]: That is fine.

**Comment by Lichtso** [comment]
Another slight change in behavior in which the on-chain version can not precisely match the original built-in is the compute meter and other resource limits. I think that should be mentioned in the document.

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:59]: For example, we may choose to make some programs upgradeable - like Feature Gate - but others could be non-upgradeable - like System - therefore they'd have separate loaders and different ways of introducing changes, if need be.

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:57]: Added.

**Comment by buffalojoec** [comment]
> I also apologize for my very late review. We need more input from the VM team and the deployment procedure is unclear. Would this be an EOA that controls the account? I think changing the address would really hurt app devs and UX.

As mentioned in [this comment](https://github.com/solana-foundation/solana-improvement-documents/pull/88#discussion_r1443124524), the address would _not_ be changed. I've added some details on this into the SIMD.

**Comment by buffalojoec** [comment]
> Another slight change in behavior in which the on-chain version can not precisely match the original built-in is the compute meter and other resource limits. I think that should be mentioned in the document.

Added.

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:79]: Up-voting previous message by @ripatel-fd. If we can move this proposal along to introduce these programs and their migration path, then each program's migration SIMD - outlining the details of its migration to BPF - can tackle privileges.

If we learn that privileged programs will need a different migration path, we can publish another SIMD that adds the necessary changes to the migration process, and block any program migration SIMDs on that new proposal.

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:17]: I think this is a good shout. Technically, many programs are implemented as native/built-in for performance or privilege reasons. It just so happens these enhanced programs are often relied on by the cluster.

The naming convention for these programs should encapsulate:
1. The program may be relied on by the cluster, but may not be
2. The program may have certain privileges, but may not
3. The program may be non-upgradeable and changed via runtime feature gates, but it may not

In my opinion:
1. "essential" (or similar) captures true, not false
2. "enshrined" (or similar) captures true, not false
3. "special" (or similar) captures true, not false

I feel like "core" is broad enough to encapsulate all branches here. I'm happy to update the proposal to relax some of the wording around "essential to cluster operations" in favor of something similar to what I've written here.

What do you think?

File: proposals/0088-enable-core-bpf-programs.md
Comment (pgarg66) [inline proposals/0088-enable-core-bpf-programs.md:116]: Dup of line 112?

Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:116]: Yes, sorry. Thanks!

**Review by pgarg66 — APPROVED** [review]

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md]: @lheeger-jump Curious what you think here. I'm leaning towards less standardization for the reasons listed above.

File: proposals/0088-enable-core-bpf-programs.md
Comment (buffalojoec) [reply proposals/0088-enable-core-bpf-programs.md:17]: @ripatel-fd If we're good with "core" I think I'll resolve this one? Lmk, thanks.

**Review by Lichtso — APPROVED** [review]

**Comment by jacobcreech** [comment]
Looks like we've got general consensus. I'll give a last call for reviews and merge on Thursday, January 25th if there are no further objections from those with level 2 access.

**Review by jacobcreech — APPROVED** [review]
LGTM. Thanks Joe for moving this along

**Review by Dakota48423 — APPROVED** [review]

**Review by Pietrek96 — APPROVED** [review]

**Review by Pietrek96 — APPROVED** [review]

**Comment by t-nelson** [comment]
can the bot org block users that approve that aren't on the approver list?

**Comment by Benhawkins18** [comment]
That could be arranged. Fixing the bot so it doesn't respond to those not on the approved list and making it ignore closed permitted pull requests

**Pull request merged**
